core-backup/src/main/scala/io/aiven/guardian/kafka/backup/BackupClientInterface.scala

```diff
@@ -503,4 +574,41 @@ object BackupClientInterface {
     // TODO handle overflow?
     ChronoUnit.MICROS.between(finalInitialTime, reducedConsumerRecord.toOffsetDateTime) / period.toMicros
   }
+
+  /** Flattens an existing flow so that for each incoming element there is exactly one outputting element. Typically
+    * this is used in combination with `unsafeDataVia` so that your `FlowWithContext`/`SourceWithContext` doesn't get
+    * corrupted.
+    * @param flow
+    *   An existing flow that you want to flatten
+    * @param zero
+    *   A zero value which is used if the `flow` doesn't produce any elements
+    * @param foldFunc
+    *   A function that determines how to concatenate a sequence
+    * @return
+    *   A flow that will always produce exactly single output element for a given input element
+    */
+  private[backup] def flattenFlow[T](flow: Flow[T, T, NotUsed], zero: T, foldFunc: (T, T) => T): Flow[T, T, NotUsed] =
```

Comment:
The context behind this function is that `unsafeDataVia` can only works if you provide a `Flow` that strictly never adds or removes elements from the stream (i.e. it can only modify each incoming element), this is the reason behind the "unsafe" in the method name. GZip compression doesn't do this, i.e. it can arbitrarily add or remove elements at will.

So one of the ways to solve this problem is to forcefully materialize the Gzip compression stream (i.e. execute the entire stream at this point in time using`runFold`) which ensures that for each incoming element, the entire element is gzip stream'ed.

Reply:
This part I didn't get, `Compression.gzip` is `ByteString` -> `ByteString`, how come we need to add or remove elements?

Reply:
So first thing is, even though we are going from `ByteString` -> `ByteString` we are dealing with a `Flow`/`Source` which means you have multiple elements of that `ByteString` (in other words its easier to think of it as Many of `ByteString` -> Many of `ByteString`). This is what the pure streaming nature of akka-streams implies, you are never dealing with a single strict element but a flow of many elements over time.

Now since we are dealing with a `FlowWithContext` we have a pairing of offset to `ByteString`, i.e.

| Commit  | Data |
| ------------- | ------------- |
| C1 | ByteString 1  |
| C2  | ByteString 2  |

The problem with GZip compression is that it does add or remove elements by design, i.e. it needs to add headers/footers at the start/end of the data stream. So that means that if we naively just use GZip compression you the following happens, firstly with `.unsafeDataVia` you have an intermediate stage that produces this.

| Commit  | Data |
| ------------- | ------------- |
| N/A | Gzipped ByteString Header  | 
| C1 | Gzipped ByteString 1  |       
| C2  | Gzipped ByteString 2  |
| N/A | Gzipped ByteString Footer  | 

After that `.unsafeDataVia` will zip up the `C1/C2` elements ignoring ones that don't have a commit causing this

| Commit  | Data |
| ------------- | ------------- |
| C1 | Gzipped ByteString 1  |       
| C2  | Gzipped ByteString 2  |

Now this is a broken/incorrect GZIp stream and it throws a `Truncated GZIP stream` which is actually exactly what is tested within akka-streams heren https://github.com/akka/akka/blob/7abc41cf4e7e8827393b181cd06c5f8ea684e696/akka-stream-tests/src/test/scala/akka/stream/io/compression/GzipWithCustomCompressionLevelSpec.scala#L27-L35.

You can also have a look at the implementation of the GZip compressor at https://github.com/akka/akka/blob/7abc41cf4e7e8827393b181cd06c5f8ea684e696/akka-stream/src/main/scala/akka/stream/impl/io/compression/GzipCompressor.scala#L24-L26, as you can see it adds a `header` and/or a `footer`.

Hope that explains everything.

Reply:
:+1: I see, got it now, thanks a lot for explanation